client_server_test.go

```diff
@@ -344,6 +344,63 @@ func TestDialTimeout(t *testing.T) {
 	}
 }
 
+// netConnDeadlineObserver fails test if read or write called without deadline.
+type netConnDeadlineObserver struct {
+	t           *testing.T
+	c           net.Conn
+	read, write bool
```

Comment:
My only qualm is these variable names are not descriptive of their purpose, making it a bit confusing to understand how this test works. Maybe rename them to `readDeadlineUsed` and `writeDeadlineUsed` ?